blog/2021/01/snoopi_deep.md

```diff
@@ -0,0 +1,221 @@
+@def title = "Profiling type-inference"
+@def authors = "Tim Holy and Nathan Daly"
+@def published = "20 January 2021"
+@def rss_pubdate = Date(2021, 1, 20)
+@def rss = """Profiling type-inference"""
+
+This is the second in a short series on what package developers can do to reduce the latency of Julia packages.
+In the [first post](https://julialang.org/blog/2021/01/precompile_tutorial/), we introduced precompilation and some of its constraints.
+We also pointed out that precompilation is closely tied to *type inference*:
+
+- precompilation allows you to cache the results of type inference, thus saving time when you start using methods defined in the package
+- caching becomes more effective when a higher proportion of calls are made with inferrable argument types (i.e., type-inference "succeeds"). Successful inference introduces links, in the form of `MethodInstance` backedges, that permit the entire call graph to be cached.
+
+As a consequence, anyone interested in precompilation needs to pay attention to type inference: how much time does it account for, where is it spending its time, and what can be done to improve caching.
+Julia itself provides the internal infrastructure needed to "spy" on inference, and the user-space utilities are in the [SnoopCompile] package.
+Starting with Julia 1.2, it became possible to do a limited amount of "spying" on inference, and this infrastructure corresponds to SnoopCompile's `@snoopi` macro.
+Julia 1.6 includes new changes that have permitted a far deeper look at what inference is doing.
+Appropriately enough, SnoopCompile calls this `@snoopi_deep`.
+
+The rich data collected by `@snoopi_deep` are useful for several different purposes.
+in this post, we'll describe the basic tool and show how it can be used to profile inference.
+Later posts will show other ways to use the data to reduce the amount of type-inference or cache its results.
+
+## Collecting the data
+
+To see `@snoopi_deep` in action, we'll use the following demo:
+
+```
+module SnoopDemo
+    struct MyType{T} x::T end
+
+    extract(y::MyType) = y.x
+
+    function domath(x)
+        y = x + x
+        return y*x + 2*x + 5
+    end
+
+    dostuff(y) = domath(extract(y))
+
+    function packintype(x)
+        y = MyType(x)
+        return dostuff(y)
+    end
```

Comment:
this name `packintype` was hard for me to parse, and i don't really know what it means.

Maybe you can call this `domathwithmytype(x)`?
Or since that's getting quite verbose, maybe `domath_with_mytype(x)`? I think i like that best.
```suggestion
    function domath_with_mytype(x)
        y = MyType(x)
        return dostuff(y)
    end
```